gcc/rust/parse/rust-parse-impl.h

```diff
@@ -12372,33 +12372,64 @@ Parser<ManagedTokenSource>::null_denotation_not_path (
       case AMP: {
 	// (single) "borrow" expression - shared (mutable) or immutable
 	std::unique_ptr<AST::Expr> expr = nullptr;
-	bool is_mut_borrow = false;
+	Mutability mutability = Mutability::Imm;
+	bool raw_borrow = false;
 
 	ParseRestrictions entered_from_unary;
 	entered_from_unary.entered_from_unary = true;
 	if (!restrictions.can_be_struct_expr)
 	  entered_from_unary.can_be_struct_expr = false;
 
-	if (lexer.peek_token ()->get_id () == MUT)
+	auto is_mutability = [] (const_TokenPtr token) {
+	  return token->get_id () == CONST || token->get_id () == MUT;
+	};
+
+	auto t = lexer.peek_token ();
+	// Weak raw keyword, we look (1) ahead and treat it as an identifier if
+	// there is no mut nor const.
+	if (t->get_id () == IDENTIFIER
+	    && t->get_str () == Values::WeakKeywords::RAW
+	    && is_mutability (lexer.peek_token (1)))
+	  {
+	    lexer.skip_token ();
+	    switch (lexer.peek_token ()->get_id ())
+	      {
+	      case MUT:
+		mutability = Mutability::Mut;
+		break;
+	      case CONST:
+		mutability = Mutability::Imm;
+		break;
+	      default:
+		rust_error_at (lexer.peek_token ()->get_locus (),
+			       "raw borrow should be either const or mut");
```

Comment:
I can't remember so feel free to disregard, but shouldn't we do something like `add_error(Error(...))` here? so that parser errors are only emitted if we want them to be?